cranelift/codegen/src/isa/riscv64/inst/emit.rs

```diff
@@ -1144,37 +1144,80 @@ impl MachInstEmit for Inst {
             &Inst::EBreak => {
                 sink.put4(0x00100073);
             }
-            &Inst::Icmp {
-                cc,
+            &Inst::I128LessThan {
                 rd,
-                ref a,
-                ref b,
-                ty,
+                ref x,
+                ref y,
+                signedness,
             } => {
-                let a = alloc_value_regs(a, &mut allocs);
-                let b = alloc_value_regs(b, &mut allocs);
+                let x = alloc_value_regs(x, &mut allocs);
+                let y = alloc_value_regs(y, &mut allocs);
                 let rd = allocs.next_writable(rd);
-                let label_true = sink.get_label();
-                let label_false = sink.get_label();
-                Inst::lower_br_icmp(
-                    cc,
-                    a,
-                    b,
-                    BranchTarget::Label(label_true),
-                    BranchTarget::Label(label_false),
-                    ty,
-                )
-                .into_iter()
-                .for_each(|i| i.emit(&[], sink, emit_info, state));
+                let is_signed = signedness.is_signed();
 
-                sink.bind_label(label_true);
-                Inst::load_imm12(rd, Imm12::TRUE).emit(&[], sink, emit_info, state);
+                // Compare the top halves of the two values. If they are equal, then
+                // we can just check the bottom halves as unsigned. Otherwise we can
+                // just check the top halves.
+                //
+                // In the unsigned case the sequence is the same, but we also compare
+                // the top halves as unsigned.
+                //
+                // Emit the following sequence:
+                //     beq       x_hi, y_hi, .top_equal
+                //     slt{,u}   rd, x_hi, y_hi
+                //     j         .end
+                // .top_equal:
+                //     sltu      rd, x_lo, y_lo
+                // .end:
```

Comment:
I guess I128 could be handled without a pseudo-instruction if you use `SelectReg` or something. Those are also pseudo-instructions hiding branches but at least they're of more general use.

That would be something like this, right? `(gen_select_reg IntCC.Equal x_hi y_hi (rv_slt x_lo y_lo) (rv_slt x_hi y_hi)`

I guess that evaluates both `slt` instructions unconditionally, so it implements I128 comparisons in a way that is slightly slower than your version. Maybe that performance benefit isn't worth having a dedicated pseudo-instruction though. What do you think?

Reply:
👍 Its a two instruction difference, I'm not too worried about i128's and it cleans up a bunch of code! All great reasons to remove it!